File: MacOS/syncFolder/Models/SyncPair.swift
import Foundation

struct SyncPair: Identifiable, Codable, Equatable, Hashable, Sendable {
    var id: UUID
    var name: String
    var sourcePath: String
    var destinationPath: String
    /// Polling interval in minutes (1-120)
    var intervalMinutes: Int
    var deleteOrphans: Bool
    var isEnabled: Bool
    var lastSyncDate: Date?
    /// Glob-like exclude patterns (e.g. "*.log", ".DS_Store", "node_modules")
    var filterPatterns: [String]
    /// Sync direction mode
    var syncMode: SyncMode

    init(name: String = "",
         sourcePath: String = "",
         destinationPath: String = "",
         intervalMinutes: Int = 5,
         deleteOrphans: Bool = false,
         isEnabled: Bool = true,
         filterPatterns: [String] = [],
         syncMode: SyncMode = .oneWay) {
        self.id = UUID()
        self.name = name
        self.sourcePath = sourcePath
        self.destinationPath = destinationPath
        self.intervalMinutes = intervalMinutes
        self.deleteOrphans = deleteOrphans
        self.isEnabled = isEnabled
        self.filterPatterns = filterPatterns
        self.syncMode = syncMode
    }

    var isValid: Bool {
        !name.trimmingCharacters(in: .whitespaces).isEmpty
            && !sourcePath.isEmpty
            && !destinationPath.isEmpty
            && sourcePath != destinationPath
            && intervalMinutes >= 1
            && intervalMinutes <= 120
    }

    var sourceURL: URL { URL(fileURLWithPath: sourcePath) }
    var destinationURL: URL { URL(fileURLWithPath: destinationPath) }
}

// MARK: - Sync Mode

enum SyncMode: String, Codable, Sendable, CaseIterable {
    case oneWay = "oneWay"
    case bidirectional = "bidirectional"
}

// MARK: - Sync Result

struct SyncResult: Codable, Equatable, Sendable {
    let filesCopied: Int
    let filesDeleted: Int
    let bytesTransferred: UInt64
    let errors: [String]

    var hasErrors: Bool { !errors.isEmpty }
    var isClean: Bool { filesCopied == 0 && filesDeleted == 0 && errors.isEmpty }

    var summary: String {
        var parts: [String] = []
        if filesCopied > 0 { parts.append(String(localized: "\(filesCopied) copied")) }
        if filesDeleted > 0 { parts.append(String(localized: "\(filesDeleted) deleted")) }
        if bytesTransferred > 0 { parts.append(ByteCountFormatter.string(fromByteCount: Int64(bytesTransferred), countStyle: .file)) }
        if parts.isEmpty { return String(localized: "No changes") }
        return parts.joined(separator: ", ")
    }
}

// MARK: - Pair Status

struct PairStatus: Equatable, Sendable {
    enum State: Equatable, Sendable {
        case idle, syncing, error
    }

    var state: State = .idle
    var errorMessage: String?
    var lastResult: SyncResult?
    var lastSyncDate: Date?
}

// MARK: - Log Entry

struct LogEntry: Identifiable, Sendable {
    let id = UUID()
    let pairId: UUID
    let pairName: String
    let result: SyncResult?
    let error: String?
    let date: Date

    init(pairId: UUID, pairName: String, result: SyncResult, date: Date) {
        self.pairId = pairId
        self.pairName = pairName
        self.result = result
        self.error = result.hasErrors ? result.errors.first : nil
        self.date = date
    }

    init(pairId: UUID, pairName: String, error: String, date: Date) {
        self.pairId = pairId
        self.pairName = pairName
        self.result = nil
        self.error = error
        self.date = date
    }

    var isError: Bool { error != nil && result == nil }

    var icon: String {
        if isError { return "exclamationmark.triangle.fill" }
        if let r = result, r.hasErrors { return "exclamationmark.circle.fill" }
        return "checkmark.circle.fill"
    }
}

File: Windows/syncFolder/Models/SyncPair.cs
using System.Text.Json.Serialization;

namespace syncFolder.Models;

public enum SyncMode
{
    [JsonPropertyName("oneWay")]
    OneWay,

    [JsonPropertyName("bidirectional")]
    Bidirectional
}

public class SyncPair
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Name { get; set; } = string.Empty;
    public string SourcePath { get; set; } = string.Empty;
    public string DestinationPath { get; set; } = string.Empty;
    public int IntervalMinutes { get; set; } = 5;
    public bool DeleteOrphans { get; set; }
    public bool IsEnabled { get; set; } = true;
    public DateTime? LastSyncDate { get; set; }
    public List<string> FilterPatterns { get; set; } = new();
    public SyncMode SyncMode { get; set; } = SyncMode.OneWay;

    [JsonIgnore]
    public bool IsValid =>
        !string.IsNullOrWhiteSpace(Name)
        && !string.IsNullOrEmpty(SourcePath)
        && !string.IsNullOrEmpty(DestinationPath)
        && !string.Equals(SourcePath, DestinationPath, StringComparison.OrdinalIgnoreCase)
        && IntervalMinutes >= 1
        && IntervalMinutes <= 120;

    public SyncPair Clone()
    {
        return new SyncPair
        {
            Id = Id,
            Name = Name,
            SourcePath = SourcePath,
            DestinationPath = DestinationPath,
            IntervalMinutes = IntervalMinutes,
            DeleteOrphans = DeleteOrphans,
            IsEnabled = IsEnabled,
            LastSyncDate = LastSyncDate,
            FilterPatterns = new List<string>(FilterPatterns),
            SyncMode = SyncMode
        };
    }
}

public class SyncResult
{
    public int FilesCopied { get; set; }
    public int FilesDeleted { get; set; }
    public ulong BytesTransferred { get; set; }
    public List<string> Errors { get; set; } = new();

    public bool HasErrors => Errors.Count > 0;
    public bool IsClean => FilesCopied == 0 && FilesDeleted == 0 && Errors.Count == 0;

    public string Summary
    {
        get
        {
            var parts = new List<string>();
            if (FilesCopied > 0) parts.Add($"{FilesCopied} copied");
            if (FilesDeleted > 0) parts.Add($"{FilesDeleted} deleted");
            if (BytesTransferred > 0) parts.Add(FormatBytes(BytesTransferred));
            if (parts.Count == 0) return Strings.Resources.NoChanges;
            return string.Join(", ", parts);
        }
    }

    private static string FormatBytes(ulong bytes)
    {
        string[] sizes = { "B", "KB", "MB", "GB", "TB" };
        double len = bytes;
        int order = 0;
        while (len >= 1024 && order < sizes.Length - 1)
        {
            order++;
            len /= 1024;
        }
        return $"{len:0.##} {sizes[order]}";
    }
}

public class PairStatus
{
    public enum StateType
    {
        Idle,
        Syncing,
        Error
    }

    public StateType State { get; set; } = StateType.Idle;
    public string? ErrorMessage { get; set; }
    public SyncResult? LastResult { get; set; }
    public DateTime? LastSyncDate { get; set; }
}
